### ISSUE TITLE
twisted.test.test_ssl.StolenTCPTestCase has a number of defects

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3306]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3306|
|Type|defect|
|Created|2008-06-23 16:23:45Z|
|Branch|https://github.com/twisted/twisted/tree/stolen-tcp-testcase-3306|

  1. it creates a server SSL context which cannot talk to the SSL client it creates.  This is because it does not supply a certificate and it does not configure EDH parameters.  Thus, no cipher can be negotiated.
  1. assuming this is fixed, it does not wait for the SSL negotiation to complete before disconnecting the client.  This is because with SSL connections, makeConnection is called before SSL negotiation happens.  Thus, the test is not a valid one for the common SSL cleanup case, as it disconnects an SSL connection which is not fully set up.
  1. assuming this is fixed, the exception OpenSSL actually produces for the final write is different from the one it expects.  This is pretty much just an implementation detail.

<details><summary>Searchable metadata</summary>

```
trac-id__3306 3306
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_stolen_tcp_testcase_3306 branches/stolen-tcp-testcase-3306
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1214238225000000 1214238225000000
changetime__1214480123000000 1214480123000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [23982]) Branching to 'stolen-tcp-testcase-3306'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [23983]) improve the ssl cleanup test a bit

  1. use PrivateCertificate to load server.pem and use it on the server, allowing the
     ssl negotiation to actually succeed
  2. write some bytes to the connection from the client before calling loseConnection,
     this forces the SSL negotiation to actually happen before the connection is dropped

refs [#3306](https://github.com/twisted/twisted/issues/3306)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [23984]) Change from one implementation detail to another implementation detail

refs [#3306](https://github.com/twisted/twisted/issues/3306)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

Well.  I changed it.  I hope it's an improvement.  The test is a bit more careful now, making sure that the connection is lost for the right reason.  Before it was being lost with an OpenSSL error, but that information was being discarded.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

The test now fails under Windows:
```
[ERROR]: twisted.test.test_ssl.StolenTCPTestCase.test_properlyCloseFiles

Traceback (most recent call last):
Failure: twisted.internet.error.ConnectionLost: Connection to the other side was lost in a non-clean fashion.
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [24025]) it doesn't really matter how the tcp connection was closed, as long as it was closed

refs [#3306](https://github.com/twisted/twisted/issues/3306)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

I changed the test to use `ConnectionClosed` instead of `ConnectionDone`.  For what the test is concerned with, this is just as good (it only cares that the connection is gone, not why or how).

I remember that Windows has some unusual behavior with respect to closed connections, perhaps only when they are running over localhost.  It seems like Twisted should be isolating users from this, but maybe that's impossible or something.  I guess if there is a problem there, it is a different ticket than this one, though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

Nice, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [24054]) Merge stolen-tcp-testcase-3306

Author: exarkun
Reviewer: therve
Fixes: [#3306](https://github.com/twisted/twisted/issues/3306)

Previously, `twisted.test.test_ssl.StolenTCPTestCase` set up an SSL server
and SSL client which could not actually talk to each other.  As such, the
test never tested that a working SSL connection would be torn down and
cleaned up correctly.  This has been fixed, and the base TCP test case has
been adjusted to assert that the connections are closed in an expected way.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
